### PR TITLE
update from infrastructure and taskmarket

### PR DESCRIPTION
**Pull request by magwas**

### CODE REVIEW

File: metamodel.zenta [hunk 1]
```diff
@@ -1,6 +1,155 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (szirbucz) [inline metamodel.zenta:1]: TestData description should be updated according to that we discussed last week

File: metamodel.zenta [hunk 2]
```diff
@@ -2044,160 +2328,659 @@ A role is assigned to human(s).
...
+            <bounds x="1064" y="584" width="-1" height="-1"/>
+            <sourceConnection xsi:type="zenta:Connection" id="97b2ef27-7968-48bc-968a-d00c6dfe5a29" source="12eeb6ff-7e88-4276-9bc3-7cdbc39a2ebf" target="9c707502-f7d0-498a-8c3c-069e3064342c" relationship="ecfdff07-9fe4-47b9-96b0-fd33d84f1fef"/>
+          </child>
+          <child xsi:type="zenta:DiagramObject" id="282f542e-4dbc-41d0-8306-c99664eabab6" targetConnections="f9f449eb-b796-4c21-95c1-3417ead121ec" zentaElement="2220484d-4957-4c8d-8f7c-041adbfb17ba">
+            <bounds x="1272" y="457" width="-1" height="-1"/>
+            <sourceConnection xsi:type="zenta:Connection" id="c0f87a00-7bcc-4f21-bf64-9c8e562debe0" source="282f542e-4dbc-41d0-8306-c99664eabab6" target="fb1f833d-e1b4-4fc5-b54b-ae378749b0c5" relationship="f177d40b-0bc4-4224-957c-0d1f9f9f66ed"/>
+          </child>
+          <child xsi:type="zenta:DiagramObject" id="fb1f833d-e1b4-4fc5-b54b-ae378749b0c5" targetConnections="c0f87a00-7bcc-4f21-bf64-9c8e562debe0" zentaElement="904ce177-b87f-42b3-90ae-53e9791e4bdb">
+            <bounds x="1293" y="589" width="-1" height="-1"/>
+
```
Comment (szirbucz) [inline metamodel.zenta:4149]: I think, review should be handled as part of the CI. It's a quality gate of the CI pipeline, even if it's done manually.
I miss the product or product version here as outcome.

**Review by szirbucz — CHANGES_REQUESTED** [review]

File: metamodel.zenta [hunk 3]
```diff
@@ -2121,8 +2109,16 @@ This is the same as Requirement in Common Criteria.
...
+            <property key="textPosition" value="1"/>
+            <property key="lineDecoration" value="SmallEndArrowDecoration"/>
+            <property key="lineColor"/>
+            <property key="fontColor"/>
+            <property key="lineWidth" value="1"/>
+            <property key="font"/>
+          </element>
         </folder>
-        <folder name="2. Process Step, Role" id="44ceef6a-62b9-434f-a3f2-fa9770958244">
+        <folder name="2. Process Step. Role" id="44ceef6a-62b9-434f-a3f2-fa9770958244">
```
Comment (szirbucz) [inline metamodel.zenta]: Was this change intentional? I think comma is better here than dot.

File: metamodel.zenta [hunk 4]
```diff
@@ -1173,7 +1059,7 @@ count(zenta:neighbours($p1,$p2,'is implemented by/implements,2')[@xsi:type='Proc
               <property key="font"/>
               <property key="documentationMarkup" value="verbatim"/>
             </element>
-            <element xsi:type="zenta:BasicObject" name="Have an output type" id="f495ac3b-b382-4e2d-9da4-98db9a03fa74" ancestor="bb3cdf1a-0bc1-4772-8913-d0567ef7ab1f">
+            <element xsi:type="zenta:BasicObject" name="have an output type" id="f495ac3b-b382-4e2d-9da4-98db9a03fa74" ancestor="bb3cdf1a-0bc1-4772-8913-d0567ef7ab1f">
```
Comment (szirbucz) [inline metamodel.zenta]: has

File: metamodel.zenta [hunk 5]
```diff
@@ -1209,25 +1095,17 @@ else true()
             <property key="Template"/>
           </element>
           <element xsi:type="zenta:BasicObject" name="Service" id="4dce3f0a-a681-447a-896b-d0ca2c1585fe" ancestor="1cbb1bec-b874-44cc-ae59-19f17e58d604">
-            <documentation>A service is a unit implementing some logic. In java it is a class and an interface. In Python it is a class, using the Java naming convention, including the one class one module rule.
+            <documentation>A service is a unit implementing some logic which is testable with unit tests.
 
-A service do not have any state. In java that means that the service only have fields for constants and the management of its dependencies (see below), no other fields. In python dependencies and constants are static: they are created in the module before the class definition.
+A service do not have any state.
```
Comment (szirbucz) [inline metamodel.zenta]: does not have

File: metamodel.zenta [hunk 6]
```diff
@@ -1413,33 +1288,8 @@ Not modeled.
...
-
-stubs describing behaviour of production code or external dependencies should be put to stub providers
-
-stubs describing behaviour of parameters should be put to test data units
-
-helper functions should be put to test helpers
-
-setup can be used to actually call the tested service, and leave only the assertions to the test cases
+            <element xsi:type="zenta:BasicObject" name="Test" id="bde80e23-f840-49c8-b167-ec430e7d9800" ancestor="thing">
+              <documentation>A unit test is a code testing (a part of) the implementation of a Service or Controller. If there is a TCDD (temporary name) framework for the given language, that should be used instead of unit tests.
```
Comment (szirbucz) [inline metamodel.zenta]: unit test also can test part of legacy classes.

File: metamodel.zenta [hunk 7]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+              <bounds x="948" y="270" width="193"/>
+            </child>
+          </element>
+          <element xsi:type="zenta:BasicObject" name="Each step have standards for its deliverables" id="58ed3b1d-6619-4674-a36c-e70b89250f2d" ancestor="e7c7a3b9-cc91-4d99-854a-d5eab68efd7c"/>
+          <element xsi:type="zenta:BasicObject" name="Standardized and simple interfaces" id="36bef80f-911f-44cb-8da1-86586c9d33e3" ancestor="e7c7a3b9-cc91-4d99-854a-d5eab68efd7c"/>
+          <element xsi:type="zenta:BasicObject" name="The process is helped by automated tools" id="2f344828-473d-4b6e-9a11-e38a63189c3e" ancestor="e7c7a3b9-cc91-4d99-854a-d5eab68efd7c"/>
+          <element xsi:type="zenta:BasicObject" name="Each step needs exactly one skillset" id="f744db66-8aa9-47d6-80dd-e1fe0d88f0d9" ancestor="e7c7a3b9-cc91-4d99-854a-d5eab68efd7c"/>
+          <element xsi:type="zenta:BasicObject" name="Implementation rules make easy to understand and continue the code" id="0a4ffc1e-7728-4dd6-a67a-211d70be292d" ancestor="e7c7a3b9-cc91-4d99-854a-d5eab68efd7c"/>
```
Comment (szirbucz) [inline metamodel.zenta:2443]: s/continue/maintain/

File: metamodel.zenta [hunk 8]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+* The most specific the use of something, the longer its name.
+* Do not be redundant.
+
+</documentation>
+          </element>
+          <element xsi:type="zenta:BasicObject" name="Focus on API" id="906e0159-1ab7-4ec0-b0a5-858519ab1f37" ancestor="behaviour">
+            <documentation>* Focus on API rather than patterns.
+* First, write down the API for perfect scenario, observe how it feels, then jump to coding and make it work.
+* If it helps understanding, extract expressions to local variables and/or methods.
+* Tell, don't ask.</documentation>
```
Comment (szirbucz) [inline metamodel.zenta]: "Tell, don't ask" should be extracted a bit. Even with example.

File: metamodel.zenta [hunk 9]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+* Similar functions should be close.
+* Place supporting functions below supported ones.
+* Use white space to associate related things and disassociate weakly related.
+                                     
+                                  </documentation>
+          </element>
+          <element xsi:type="zenta:BasicObject" name="KISS" id="a554b350-a076-44e8-96bb-083ca197c1fd" ancestor="behaviour">
+            <documentation>Keep it simple, stupid.
+* Extract the complex logic to other method.
+* Avoid the use of else.
```
Comment (szirbucz) [inline metamodel.zenta]: Overusing if-else can cause high Cyclomatic complexity, but it's not the root of all evil. Sometimes it helps to keep your code readable.

File: metamodel.zenta [hunk 10]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+                                     
+                                  </documentation>
+          </element>
+          <element xsi:type="zenta:BasicObject" name="Test independency" id="50f8531a-d8da-4360-99b7-b3c899e7c07a" ancestor="behaviour">
+            <documentation>* The tests should organize actions on actual states.
+* Do not wait a specific amount of time:
+** use callbacks instead of polling
+** if the only way is polling, set the poll frequency low, and time out after a reasonable number of attempts.
+*  If the test needs an interface for observing the actual state then create an interface for it.
+* If the behaviour is time-dependent, then dependency inject the clock
```
Comment (szirbucz) [inline metamodel.zenta]: A mocked clock

File: metamodel.zenta [hunk 11]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+          </element>
+          <element xsi:type="zenta:BasicObject" name="CI is for CI" id="4c79a2cf-4a32-4290-a359-76f112f207f3" ancestor="behaviour">
+            <documentation>CI mechanisms should be used only for CI
+using the CI mechanisms for gaining unauthorized access, or for any other purpose than building and quality checking the project is strictly forbidden
+</documentation>
+          </element>
+          <element xsi:type="zenta:BasicObject" name="New dependencies" id="c0de6c9e-de2c-46d1-85f4-4e4706a8da6c" ancestor="behaviour">
+            <documentation>* New external dependencies should not be added lightly
+* While writing functionality someone else have already written is waste of effort, a dependency should be good quality and with appropriate licence.
+* Introducing a new dependency should only be done with the approval of the designer.
```
Comment (szirbucz) [inline metamodel.zenta:2744]: I'm wondering if designers will be skilled enough to make this decision. It might be the responsibility of the architect.

File: metamodel.zenta [hunk 12]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+          <element xsi:type="zenta:BasicObject" name="git-flow" id="1bab363b-0980-48ca-9506-d1b07bed44f2" ancestor="behaviour">
+            <documentation>Use git-flow. See https://nvie.com/posts/a-successful-git-branching-model/
+                                     
+* The code should be based in latest development branch, and the pull request should go into the same. Create feature branches for all your tasks.
+</documentation>
+          </element>
+          <element xsi:type="zenta:BasicObject" name="grammar-based naming conventions" id="519bd2fd-3152-496d-8895-23925fea3614" ancestor="behaviour">
+            <documentation>All names are descriptive.
+
+* module, unit, parameter, field,  and local variable names are nouns
```
Comment (szirbucz) [inline metamodel.zenta]: double space after "field,"

File: metamodel.zenta [hunk 13]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+* code unit names refer to the activity/process they support
+* method names are verbs, except
```
Comment (szirbucz) [inline metamodel.zenta]: except what? the following? Why are they at the same level in the list then?

File: metamodel.zenta [hunk 14]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+* the name of a method creating a complex constant should start with 'get'
+* the name of the method creating a collection of constants should start with the type of the collection, e.g. 'list'
+* the name of the method giving back a default instance ( a fully built instance meant to be used in the positive test case) is 'get()'
+* the name of the method giving back a collection containing only the default instance should be the name of the collection type, e.g. 'list()'
+* other methods must directly or indirectly use the get() method to deliver other instances by modifying the default instance.
+* the non-default methods must contain the name of the (most important) field modified right after the 'get' part. For example the method giving back an Entity with a zero id would be called getIdUninitialized()</documentation>
+            </element>
+            <element xsi:type="zenta:BasicObject" name="constant provider" id="861fcab2-587d-45f1-8587-1979180ee4f6" ancestor="behaviour">
+              <documentation>* constants can only be in constant provider units; move them there from other units
```
Comment (szirbucz) [inline metamodel.zenta:2803]: As far as I remember we've been talking about constants used in a single place can be defined locally. Am I wrong?

File: metamodel.zenta [hunk 15]
```diff
@@ -2335,180 +2365,1111 @@ A role is assigned to human(s).
...
+              <documentation>* Units without naming convention (Delegate, Configuration provider, and units which part of the boilerplate needed with the frameworks used) should have an @InterfaceClass annotation, denoting the lack of naming convention applicable.</documentation>
+            </element>
+            <element xsi:type="zenta:BasicObject" name="modeled placement" id="6a6a5f4e-be42-4cad-bf72-be0116b8817e" ancestor="behaviour">
+              <documentation>First-class production code units are placed in the production code hierarchy under the package designated by the design
+</documentation>
+            </element>
+            <element xsi:type="zenta:BasicObject" name="No reference to production code constants" id="5397f74e-8346-4361-a715-1118727df296" ancestor="behaviour">
+              <documentation>Test units must not reference constants in production code in any way. If a constant is needed, it should be provided by test data.</documentation>
+            </element>
+
```
Comment (szirbucz) [inline metamodel.zenta:3061]: What about deleteBy, countBy, streamBy, existBy and save?

Comment (szirbucz) [inline metamodel.zenta:3110]: What about languages where no display name feature exists? Wouldn't it be simpler to use a descriptive test method name?

**Review by szirbucz — CHANGES_REQUESTED** [review]

File: metamodel.zenta (hunk 1 above)
Comment (magwas) [reply metamodel.zenta:1]: I beleive I have done it

File: metamodel.zenta (hunk 11 above)
Comment (magwas) [reply metamodel.zenta:2744]: Let's discuss this part. What an analyst and a designer does, and how we imagine the responsibility split between them.

File: metamodel.zenta (hunk 14 above)
Comment (magwas) [reply metamodel.zenta:2803]: It have impacts on code checks. Let's discuss.

File: metamodel.zenta (hunk 15 above)
Comment (magwas) [reply metamodel.zenta:3110]: We can add an annotation or other language construct supported by the language. In worst case it can be a specially formatted comment. With function names we don't have the flexibility for punctuations and such.
For javascript consider this QUnit example, which is okay for me:

```
test("This is the name for this test", function () { 
    ok(true);	
});
```

File: metamodel.zenta (hunk 2 above)
Comment (magwas) [reply metamodel.zenta:4149]: For that we should model the management workflow which we are not sure enough of.

**Review by szirbucz — APPROVED** [review]
